static/js/stream_list.js

```diff
@@ -121,63 +103,40 @@ exports.build_stream_list = function () {
         return;
     }
 
-    streams = filter_streams_by_search(streams);
+    // The main logic to build the list is in stream_sort.js, and
+    // we get thre lists of streams (pinned/normal/dormant).
+    var stream_groups = stream_sort.sort_groups(get_search_term());
+
+    if (stream_groups.same_as_before) {
+        return;
+    }
```

Comment:
I followed your suggestion to just do a complete redraw of the stream list when we remove a row.  We already had a bug in the past with the incremental approach--it was fixed, but in an ugly way, and then I was about to re-break it.

I think moving to an consistent approach of always re-building the list on add/remove will make the `same_as_before` logic valid.  I'm not crazy about those checks--I wonder if there is a better way to short circuit updates further upstream--but we've had them in place for a while, and this is just essentially moving existing code.